### ISSUE TITLE
How to ingect a customized population to the problem to run MOEA not a random initialization?

### ISSUE THREAD

**Issue by fchgithub** [MOEAFramework/MOEAFramework#162]
Hi, 
In my problem, I need to collect the population from one run and load that population to another run (maybe to run with another algorithm). In fact, I want to avoid random initialization and use my own population to start another round of MOEA. How is it possible?

Any help is appreciated (this is an urgent need)

Thanks,

Fatemeh

**Comment by dhadka** [comment]
Hi, thanks for the question.  You essentially need to use the `InjectedInitialization` class to initialize the algorithm with the solutions from the previous run.

The first step would be to run an algorithm and get its final population:

```java

        Initialization initialization = new RandomInitialization(
                problem,
                100);

        TournamentSelection selection = new TournamentSelection(2, 
                new ChainedComparator(
                        new ParetoDominanceComparator(),
                        new CrowdingComparator()));

        Variation variation = new GAVariation(
                new SBX(1.0, 25.0),
                new PM(1.0 / problem.getNumberOfVariables(), 30.0));

        NSGAII algorithm = new NSGAII(
                problem,
                new NondominatedSortingPopulation(),
                null, // no archive
                selection,
                variation,
                initialization);

        while (algorithm.getNumberOfEvaluations() < 10000) {
            algorithm.step();
        }

        Population intermediateResult = algorithm.getPopulation();
```

Then inject those solutions into the second algorithm:

```java

        initialization = new InjectedInitialization(
                problem,
                100,
                intermediateResult);

        algorithm = new NSGAII(
                problem,
                new NondominatedSortingPopulation(),
                null, // no archive
                selection,
                variation,
                initialization);

        while (algorithm.getNumberOfEvaluations() < 10000) {
            algorithm.step();
        }

        NondominatedPopulation finalResult = algorithm.getResult();
```

**Comment by fchgithub** [comment]
Awesome, thanks!
I just want to make sure if `algorithm.getPopulation()` would return the current state of the population regardless of nondomination rule? I need all 100 individuals whatever they are!

**Comment by dhadka** [comment]
Yes, getPopulation() returns the entire population.  getResult() only returns the non-dominated solutions.

**Comment by fchgithub** [comment]
Perfect! thank you so much for the quick response.

**Comment by fchgithub** [comment]
Hi again,

I am trying to use NSGAIII and GDE3 as well (I looked at StandardAlgorithms class for the required parameters), but it keeps getting errors. For example, in GDE3 

`DominanceComparator comparator = new ParetoDominanceComparator();

		NondominatedSortingPopulation population = new NondominatedSortingPopulation(comparator);`  

gives me typecast error while they are working fine in StandardAlgorithm class. 

Regarding using NSGAIII, I am not sure how I can create one in the same way as NSGAII. 

Thanks,
Fatemeh

**Comment by dhadka** [comment]
Please check your imports and make sure you're not using one of jmetal's files.  There are a few classes with identical names.  For example, for DominanceComparator, you could be importing

    import jmetal.util.comparators.DominanceComparator;

whereas you really want:

    import org.moeaframework.core.comparator.DominanceComparator;

If this doesn't help, could you please include the entire file or a small code sample that reproduces the error?

**Comment by fchgithub** [comment]
Perfect, yes it was jmetal. Thanks, DGE3 works now!
How about using NSGAIII? I assume I should copy all the code in newNSGAIII in Standard class, right? But, what should be  "properties" here which is needed to be passed to Variation?

**Comment by dhadka** [comment]
Yes, just copy the code in newNSGAIII and modify as needed.  You can pass in an empty properties object (`new Properties()`) to just use the defaults.  It will pick the default Variation class(es) based on the type of decision variables.

If you want to use a different operator than the defaults, then you can specify the "operator" property.  For example:

```java

    Properties properties = new Properties();
    properties.setProperty("operator", "pcx+pm"); // Parent centric crossover with polynomial mutation
    properties.setProperty("pcx.parents", "5"); // Use 5 parents
    properties.setProperty("pcx.offspring", "2"); // Generate 2 offspring

    Variation variation = OperatorFactory.getInstance().getVariation(null, properties, problem);
```

The properties available for each operator can be found here: [Operator Properties](http://moeaframework.org/javadoc/org/moeaframework/core/operator/StandardOperators.html)

Since you're creating the algorithm directly, you may also find it useful to just construct the variation operator directly instead of using the OperatorFactory.  For example:

```java

    Variation variation = new GAVariation(
        new SBX(1.0, 15.0),
        new PM(1.0 / problem.getNumberOfVariables(), 20.0));
```

**Comment by fchgithub** [comment]
Thanks so much for your help! It worked.

**Comment by fchgithub** [comment]
Hi @dhadka,
As my implementation progresses, I see something as a result that I cannot find out the reason for this behavior. In the following image, you can see the solutions from the whole final population:

![population-5000-run](https://user-images.githubusercontent.com/10403841/38261190-886dd1e4-3737-11e8-9298-70de214dea42.png)

They are stored in a Population object. When I add them to a NonDominatedPopulation object, only one solution is left there which even by its value shouldn't be a nondominated solution! 

![nondominatedsolutions-5000-run](https://user-images.githubusercontent.com/10403841/38261381-0afc51e4-3738-11e8-91aa-c4a1fc49c4d5.png)
 
Do you have any comment for such a wrong (in my opinion) behavior?

Thanks,
Fatemeh

**Comment by dhadka** [comment]
How many objectives does your problem have?

**Comment by fchgithub** [comment]
I am testing it with two objectives for now. I should add the third objective later, after I am sure everything works fine!

**Comment by dhadka** [comment]
Oh, I think I see the issue.  You'll want to evaluate the solutions **before** adding them to the NondominatedPopulation.

```
myProblem problem = new myProblem();
Population allSolutions = new Population();
//shortened the detail of the code here which reads the solutions from an available list by creating a Solution object and add the values to it
for (Solution solution : mySolutionList){
   allSolutions.add(solution);
}

for (Solution solution : allSolutions){
    problem.evaluate(solution );
}

NondominatedPopulation mergedSolutions = new NondominatedPopulation(allSolutions);
```

**Comment by dhadka** [comment]
Just some additional info in case others encounter this or similar issues...

When calling `new double[N]` in Java, the array is initialized with default values of 0.0.  So when calling `problem.newSolution()` or `new Solution(...)`, the objectives and constraints are initialized to 0.0.  Thus, it is necessary to always evaluate solutions before performing an operation that reads the objectives or constraints.

By default, `NondominatedPopulation` does not allow solutions with duplicate objective values.  So in your example where the solutions have not yet been evaluated and their objective values are 0.0, it removes all but one of the solutions.  You can change this behavior by using the constructor `new NondominatedPopulation(DuplicateMode.ALLOW_DUPLICATE_OBJECTIVES)` or `new NondominatedPopulation(DuplicateMode.ALLOW_DUPLICATES)`.  However, I would recommend the first solution of always evaluating solutions before adding them to the `NondominatedPopulation`.

**Comment by fchgithub** [comment]
True point! it was hard to catch, Thanks for the correction!
This is how it looks like now:

![nondominatedsolutions-500-run](https://user-images.githubusercontent.com/10403841/38265970-7046b460-3745-11e8-8204-1d70e3547273.png)

**Comment by fchgithub** [comment]
Hi @dhadka ,
A question regarding an answer in the issue #26 (with small modifications):
```
Analyzer analyzer = new Analyzer()
    .withProblem(problem)
    .includeHypervolume()
    .showStatisticalSignificance();

analyzer.addAll("GDE3" ,  executor.withAlgorithm(algorithm).run());


Population myAlgorithmSolutions = loadMySolutions(i);

analyzer.add("MyAlgorithm", new NondominatedPopulation(myAlgorithmSolutions ));


analyzer.printAnalysis();
```
Assuming there is no reference set available for the problem, 
I- are the two Pareto fronts are compared against one reference point? Becuase if each HV is calculated with it's own reference point (using min and max of its objective values), then comparing two HV doesn't make sense!
II- if the answer is yes, how I can find  out what is that reference point and how it is found?
III- in general how valuable it is to compare to Pareto fronts in this way? is there another (probably a better way to find out which algorithm is doing better?

**Comment by dhadka** [comment]
1. If no explicit reference set is provided, the Analyzer combines all of the Pareto fronts you have added to the Analyzer to create an approximation of the reference set.  The reference point is then computed from this reference set.  So the answer is yes, it's comparing the two Pareto fronts against one reference point.

2. The reference point defaults to (1, 1, ..., 1).  This is because the hypervolume calculations are normalized (the same is true for most performance indicators).  The objective values of each Pareto front are scaled to lie within [0, 1], using the reference set to determine the min/max bounds for scaling.  You can get the reference set by calling `analyzer.getReferenceSet()`.  You can also optionally add a "delta" offset to the reference point (see the settings in `global.properties`).

3. Using these performance indicators is the standard way for comparing MOEAs.  However, you shouldn't rely on a single indicator like hypervolume.  A Pareto front can exhibit different traits, such as convergence or diversity, and different performance indicators measure these traits differently.  For example, a Pareto front comprised of a single point that lies somewhere on the **true** Pareto front will have a generational distance of 0, the best possible value, but would be comparatively worse in other indicators, such as inverted generational distance and hypervolume.

**Comment by fchgithub** [comment]
Thanks a lot for the clear explanation! My assumption is that for the other performance metrics (generational distance, inverted generational distance, ...) the same reference set found by the analyzer is used, am I right? It just doesn't make sense to me in the case of the existence of only reference point, not a set!
Returning back to a basic question, I noticed in the final nondominated solution (Pareto front)  there are infeasible solutions. I didn't get why they are there! and are they playing a role to dominate other individuals in the population?

Thanks!

**Comment by dhadka** [comment]
When there are constraints, comparing or selecting the preferred solution with Pareto dominance works as follows:

1. If both solutions violate constraints, the solution with the smaller constraint violation is preferred
2. If one solution is feasible and the other violates constraints, the feasible solution is preferred
3. If both solutions are feasible, apply Pareto dominance.  One of these conditions will occur:
    1. If A dominates B, prefer A.
    2. If B dominates A, prefer B.
    3. If A and B do not dominate one another (i.e., are non-dominated), the both are equally preferred.

So if you see infeasible solutions in the final nondominated population, it typically means no feasible solutions were found, since if it did find a feasible solution, the feasible solution would dominate all infeasible solutions.

I suspect you'll find all solutions in your final nondominated population are infeasible.  *If that's not the case and you see both feasible and infeasible solutions in the final nondominated population, that's an issue I'd need to look at.*

To answer your second question "are they playing a role to dominate other individuals in the population?"...yes, infeasible solutions dominate all other infeasible solutions with worse constraint violations.

For more details, see
> Deb, K, "An Efficient Constraint Handling Method for Genetic Algorithms." Computer Methods in Applied Mechanics and Engineering, pp. 311--338, 1998.

**Comment by fchgithub** [comment]
Thanks, clear enough! I should revisit the problem! It's a very hard problem and I see just one infeasible solution.

**Comment by fchgithub** [comment]
Hi @dhadka,

In early answers in this issue, there was a discussion of using InjectedInitialization. As you nicely showed in a sample code, 
would 
`Population intermediateResult = algorithm.getPopulation();`
and then
```initialization = new InjectedInitialization(problem, 100, intermediateResult);```
bring the values of objectives to the second run? 
I need to know if the second algorithm can use those in the selection, in case the second algorithm uses a comparator for the selection. I want somehow benefit from that information in early iterations of the second algorithm. 
I hope I was clear enough to convey my question!

**Comment by dhadka** [comment]
Yes.  The second run will be initialized with the solutions from the results of the first run.

**Comment by fchgithub** [comment]
Hi @dhadka,

Another issue is that I want to add a repair function which gets the solution in the "evaluate" function and if it violates the constraints (with some probability) it repairs the values in the solution.
There is a `solution.setVariable(int index, Variable variable)` , but the only way I found to create a variable (my problem works with real values) is using `EncodingUtils.newReal(..)` which accepts lower and upper bounds to generate a random value in the range, but I want to add the value came from repair method! 
I tried to set lower and upper bounds in this `newReal` to the same value (my repaired value) but it sets the solution variable to NAN!

Is there a reason behind this that we cannot change the values of the variables is a solution? How can I force it to the solution? 

Thanks,

Fatemeh
